### PR TITLE
LLVM and SPIRV-LLVM-Translator pulldown (WW41)

### PR DESCRIPTION
**Pull request by sys-ce-bb**
LLVM: llvm/llvm-project@894927b491b7
SPIRV-LLVM-Translator: KhronosGroup/SPIRV-LLVM-Translator@e68ddf71512a77a6bae4b281578d52e0ab98ab2f

### CODE REVIEW

**Comment by jsji** [comment]
https://github.com/intel/llvm/actions/runs/6503128621/job/17663436008?pr=11521

4 matrix tests are timeout.

Timed Out Tests (4):
  SYCL :: Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp
  SYCL :: Matrix/XMX8/element_wise_all_ops.cpp
  SYCL :: Matrix/XMX8/element_wise_all_ops_int8.cpp
  SYCL :: Matrix/XMX8/element_wise_all_ops_int8_packed.cpp


Had a quick look at element_wise_all_ops_int8_packed.cpp, looks like some loads are not optimized out due to new commits from llorg now?
@YuriPlyakhin  @dkhaldi Are you able to help to have a look? Or do you want me to disable these tests for now and open issues? Thanks.

```

 ; Function Attrs: convergent norecurse nounwind
-define weak_odr dso_local spir_kernel void @_ZTSZZ17matrix_verify_mulIaLm16ELm16EEvN4sycl3_V15queueER10big_matrixIT_XT0_EXT1_EERNS1_8nd_rangeILi2EEEiENKUlRNS1_7handlerEE_clESB_E10mul_matrix(ptr addrspace(1) noundef align 1 %_arg_accA) local_unnamed_addr #0 comdat !srcloc !95 !kernel_arg_buffer_location !62 !kernel_arg_runtime_aligned !63 !kernel_arg_exclusive_ptr !63 !intel_reqd_sub_group_size !64 !sycl_fixed_targets !65 !sycl_kernel_omit_args !66 {
+define weak_odr dso_local spir_kernel void @_ZTSZZ17matrix_verify_mulIaLm16ELm16EEvN4sycl3_V15queueER10big_matrixIT_XT0_EXT1_EERNS1_8nd_rangeILi2EEEiENKUlRNS1_7handlerEE_clESB_E10mul_matrix(ptr addrspace(1) noundef align 1 %_arg_accA, ptr noundef byval(%"class.sycl::_V1::range") align 8 %_arg_accA2, ptr noundef byval(%"class.sycl::_V1::id") align 8 %_arg_accA3) local_unnamed_addr #0 comdat !srcloc !95 !kernel_arg_buffer_location !62 !kernel_arg_runtime_aligned !63 !kernel_arg_exclusive_ptr !63 !intel_reqd_sub_group_size !64 !sycl_fixed_targets !65 !sycl_kernel_omit_args !66 {
 entry:
   call spir_func void @__itt_offload_wi_start_wrapper()
+  %agg.tmp5.sroa.0.sroa.2.0._arg_accA2.ascast.sroa_idx = getelementptr inbounds i8, ptr %_arg_accA2, i64 8
+  %agg.tmp5.sroa.0.sroa.2.0.copyload = load i64, ptr %agg.tmp5.sroa.0.sroa.2.0._arg_accA2.ascast.sroa_idx, align 8
+  %agg.tmp6.sroa.0.sroa.0.0.copyload = load i64, ptr %_arg_accA3, align 8
+  %agg.tmp6.sroa.0.sroa.2.0._arg_accA3.ascast.sroa_idx = getelementptr inbounds i8, ptr %_arg_accA3, i64 8
+  %agg.tmp6.sroa.0.sroa.2.0.copyload = load i64, ptr %agg.tmp6.sroa.0.sroa.2.0._arg_accA3.ascast.sroa_idx, align 8
   %0 = load i64, ptr addrspace(1) getelementptr inbounds (i8, ptr addrspace(1) @__spirv_BuiltInGlobalInvocationId, i64 8), align 8, !noalias !96
   %1 = load i64, ptr addrspace(1) @__spirv_BuiltInGlobalInvocationId, align 32, !noalias !96
   %2 = load i64, ptr addrspace(1) getelementptr inbounds (i8, ptr addrspace(1) @__spirv_BuiltInLocalInvocationId, i64 8), align 8, !noalias !103
@@ -163,10 +194,16 @@
   br label %for.cond.i.i, !llvm.loop !108

 _ZZZ17matrix_verify_mulIaLm16ELm16EEvN4sycl3_V15queueER10big_matrixIT_XT0_EXT1_EERNS1_8nd_rangeILi2EEEiENKUlRNS1_7handlerEE_clESB_ENKUlNS1_7nd_itemILi2EEEE_clESE_.exit: ; preds = %for.cond.i.i
+  %mul.i6.i.i.i.i = mul i64 %agg.tmp6.sroa.0.sroa.0.0.copyload, %agg.tmp5.sroa.0.sroa.2.0.copyload
+  %4 = getelementptr i8, ptr addrspace(1) %_arg_accA, i64 %mul.i6.i.i.i.i
+  %add.ptr.i = getelementptr i8, ptr addrspace(1) %4, i64 %agg.tmp6.sroa.0.sroa.2.0.copyload
   %sub5.i = sub nsw i64 %1, %3
   %sub.i = sub nsw i64 %0, %2
+  %add.i7.i.i.i.i.i.i = add i64 %mul.i6.i.i.i.i, %agg.tmp6.sroa.0.sroa.2.0.copyload
+  %idx.neg.i.i.i = sub i64 0, %add.i7.i.i.i.i.i.i
+  %add.ptr.i.i.i = getelementptr inbounds i8, ptr addrspace(1) %add.ptr.i, i64 %idx.neg.i.i.i
   %mul13.i = shl nsw i64 %sub.i, 9
-  %add.ptr.i.i = getelementptr inbounds i8, ptr addrspace(1) %_arg_accA, i64 %mul13.i
+  %add.ptr.i.i = getelementptr inbounds i8, ptr addrspace(1) %add.ptr.i.i.i, i64 %mul13.i
   %div16.i = shl nsw i64 %sub5.i, 2
   %mul15.i = and i64 %div16.i, -32
   %add.ptr.i30.i = getelementptr inbounds i8, ptr addrspace(1) %add.ptr.i.i, i64 %mul15.i
@@ -176,9 +213,14 @@
 }

```

**Comment by jsji** [comment]
- [ ] [Revert "[PATCH] [llvm] [InstCombine] Canonicalise ADD+GEP"](https://github.com/intel/llvm/pull/11521/commits/e23b203aa79cd3af2d918fa8f2410018a03b34d4) This is something that we should follow up to improve . https://github.com/intel/llvm/issues/11538 opened to track.

**Comment by bader** [comment]
I really wish we don't revert InstCombine changes as they wide impact on many compilers/modes.
Can we keep the change and disable 4 failing tests instead?

File: sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 //
 //===----------------------------------------------------------------------===//
 // REQUIRES: matrix-xmx8
+// XFAIL:*
```
Comment (bader) [inline sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp:9]: Does it help?
You said in the tracker that tests are timing out, so probably it's better to skip running them at all with:

// REQUIRES: TEMPORARY_DISBLED

Comment (jsji) [reply sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp:9]: Yes, these are passing now. https://github.com/intel/llvm/actions/runs/6513111676/job/17692457598?pr=11521

This is similar to how others tests are disabled. 

> Unsupported Tests (112):
>   SYCL :: Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp
> ...

Comment (bader) [reply sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp:9]: > This is similar to how others tests are disabled.

What tests do you refer to?
NOTE: they might be disable for a different reason - most likely they fail execution within the time limit. Using XFAIL in this case is reasonable way to fix the return status.

> 
> > Unsupported Tests (112):
> > SYCL :: Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp
> > ...

> Yes, these are passing now. https://github.com/intel/llvm/actions/runs/6513111676/job/17692457598?pr=11521

According to my understanding using XFAIL can't move the test to "Unsupported Tests" category. I think what plays role here is: "REQUIRES: matrix-xmx8". The link your provided is for the run on GPU device only and my guess is that none of GHA CI systems have this feature. 

Here is the link where tests are executed on CPU as well: https://github.com/intel/llvm/actions/runs/6513111676/job/17692457598?pr=11521

NOTE: that GHA CI machines are different. I think we were lucky, and tests were executed on the machine w/o matrix support, so tests skipped. 

From https://www.llvm.org/docs/TestingGuide.html#tips-for-writing-constraints

> ``UNSUPPORTED`` and ``XFAIL``

> Both of these indicate that the test isn’t expected to work; however, they have different effects. **UNSUPPORTED causes the test to be skipped; this saves execution time**, but then you’ll never know whether the test actually would start working. Conversely, **XFAIL actually runs the test but expects a failure output, taking extra execution time** but alerting you if/when the test begins to behave correctly (an XPASS test result). You need to decide which is more appropriate in each case.

I suggest replacing `XFAIL:*` with `REQUIRES: TEMPORARY_DISBLED` to disable the test on all machines.

Comment (bader) [reply sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp:9]: @aelovikov-intel, @YuriPlyakhin, can you confirm that GHA CI have machines with and w/o `matrix-xmx8` feature support and E2E tests can be randomly scheduled on any of these machines?

Comment (YuriPlyakhin) [reply sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp:9]: @bader, GHA CI has 1-2 systems, which can run matrix-xmx8 tests, and matrix-xmx8 tests are scheduled (or will be scheduled) only on these machines. matrix-xmx8 tests are not running on other machines in GHA CI.

Comment (bader) [reply sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp:9]: Talked to @aelovikov-intel and he told me that pre-commit runs tests only in multi-device environment, so `XFAIL:*` effectively disables these tests, but we run on a single device in post-commit, where will run executed.

Comment (jsji) [reply sycl/test-e2e/Matrix/Legacy/XMX8/element_wise_all_ops_int8.cpp]: Thanks @bader @aelovikov-intel and @YuriPlyakhin . Updated to use `REQUIRES: TEMPORARY_DISBLED` instead.

**Review by bb-sycl — APPROVED** [review]

**Comment by bb-sycl** [comment]
Sat 14 Oct 2023 04:23:39 PM UTC --- Start to merge the commit into sycl branch. It will take several minutes.

**Comment by bb-sycl** [comment]
Sat 14 Oct 2023 04:27:38 PM UTC --- Merge the branch in this PR to base automatically. Will close the PR later.

**Pull request merged**